src/sinks/mqtt/sink.rs

```diff
@@ -0,0 +1,211 @@
+use std::{
+    fmt::Debug,
+};
+
+use async_trait::async_trait;
+use bytes::BytesMut;
+use futures::{
+    pin_mut,
+    stream::BoxStream,
+    Stream, StreamExt,
+};
+use rumqttc::{
+    AsyncClient, ClientError, ConnectionError,
+    EventLoop, MqttOptions,
+    QoS,
+};
+use snafu::{ResultExt, Snafu};
+use tokio_util::codec::Encoder as _;
+use vector_core::{
+    internal_event::{BytesSent, EventsSent},
+    ByteSizeOf,
+};
+
+use crate::{
+    codecs::{Encoder, Transformer},
+    emit,
+    event::{Event, EventStatus, Finalizable},
+    internal_events::{
+        ConnectionOpen, OpenGauge, MqttClientError, MqttConnectionError,
+    },
+    internal_events::TemplateRenderingError,
+    sinks::util::StreamSink,
+    sinks::mqtt::config::MqttSinkConfig,
+    template::{Template, TemplateParseError},
+    tls::TlsError,
+};
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub))]
+pub enum MqttError {
+    #[snafu(display("invalid topic template: {}", source))]
+    TopicTemplate { source: TemplateParseError },
+    #[snafu(display("MQTT connection: {}", source))]
+    Connection { source: ConnectionError },
+    #[snafu(display("TLS error: {}", source))]
+    Tls { source: TlsError },
+    #[snafu(display("MQTT client: {}", source))]
+    Client { source: ClientError },
+}
+
+#[derive(Clone)]
+pub struct MqttConnector {
+    options: MqttOptions,
+    topic: Template,
+}
+
+impl MqttConnector {
+    pub fn new(options: MqttOptions, topic: String) -> Result<Self, MqttError> {
+        let topic = Template::try_from(topic).context(TopicTemplateSnafu)?;
+        Ok(Self {
+            options,
+            topic,
+        })
+    }
+
+    fn connect(&self) -> (AsyncClient, EventLoop) {
+        AsyncClient::new(self.options.clone(), 1024)
+    }
+
+    pub async fn healthcheck(&self) -> crate::Result<()> {
+        let (client, connection) = self.connect();
+        drop(client);
+        drop(connection);
+        Ok(())
+    }
+}
+
+pub struct MqttSink {
+    transformer: Transformer,
+    encoder: Encoder<()>,
+    connector: MqttConnector,
+}
+
+impl MqttSink {
+    pub fn new(
+        config: &MqttSinkConfig,
+        connector: MqttConnector,
+    ) -> crate::Result<Self> {
+        let transformer = config.encoding.transformer();
+        let serializer = config.encoding.build()?;
+        let encoder = Encoder::<()>::new(serializer);
+
+        Ok(Self {
+            transformer,
+            encoder,
+            connector,
+        })
+    }
+
+    async fn handle_events<I>(
```

Comment:
This function has enough logic it would be good to add some inline comments to describe what is going on and/or a function comment.